### PR TITLE
Add Jupyter notebook to demonstrate query-time NDD.

### PR DESCRIPTION
**Pull request by alexaryn**
This is meant to be referenced by the NDD blog post.

### CODE REVIEW

File: notebooks/ndd_example.ipynb [hunk 1]
```diff
@@ -0,0 +1,202 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "ed9dd0a2-6b12-436f-b6c3-02e0072115e3",
+   "metadata": {},
+   "source": [
+    "This notebook shows the results of query-time near-duplicate detection (NDD).  It shows the same query with and without the duplicates.  To use this notebook, first start the Sycamore containers using `docker compose up`.  Then, ingest the college credit card marketing agreement database.  There are two choices here.  The quickest approach is to ingest only the 35 documents that are needed for the query below.  This can be accomplished by running `docker compose run sycamore_crawler_s3 -anon aryn-public cccmad-tiny`.  The other option is to ingest all 1911 PDFs that have text, by running `docker compose run sycamore_crawler_s3 -anon aryn-public cccmad`.  This will take hours and around 20 PDFs will have problems.  Once ingestion has sufficiently completed, run this notebook."
```
Comment (jonfritz) [inline notebooks/ndd_example.ipynb:8]: "Taking hours and 20 PDFs having problems" - what problems are there? Does it fail the job, or just not show up in query results. Also, how long did it take on your MacBook (or whatever we tested)? Do we know why it takes so long to do this?

File: notebooks/ndd_example.ipynb [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    "                    },\n",
+    "                ],\n",
+    "            },\n",
+    "        },\n",
+    "    }\n",
+    "    with requests.get('https://opensearch:9200/_plugins/_ml/models/_search', json=query, verify=False) as resp:\n",
+    "        res = json.loads(resp.text)\n",
+    "        return res['hits']['hits'][0]['_id']"
+   ]
+  },
```
Comment (jonfritz) [inline notebooks/ndd_example.ipynb:51]: Is it worth putting a note here explaining why we are doing this? We could link to the section of the docs that discusses this from the note, too.

File: notebooks/ndd_example.ipynb [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    "        answer = res['ext']['retrieval_augmented_generation']['answer']\n",
+    "        print(f'[ANSWER]\\n{answer}')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "61a71039-0a85-4e14-b42a-0969d04d979a",
+   "metadata": {},
+   "source": [
+    "First, we run the query without near-duplicate-detection.  We do this by not asking for `shingles` in `_source`.  The results we get have many copies of the first chunk.  The generated answer comes from only one chunk."
```
Comment (jonfritz) [inline notebooks/ndd_example.ipynb:77]: I don't think we use this syntax anywhere in examples or our docs for OpenSearch queries. You might need to explain what 'source' is in the query.

File: notebooks/ndd_example.ipynb [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+      "The rules of arbitration typically involve a dispute being settled by an arbitrator in accordance with the commercial Arbitration Rules of an organization like the American Arbitration Association [1]. The arbitrator's decision is binding and can be enforced in any court of competent jurisdiction [1]. If an arbitrator is unable or unwilling to proceed, they can be replaced according to the rules of the relevant organization [2]. The arbitration process aims to be efficient and expeditious, with the arbitrator determining the scope of discovery and other procedures [4]. The arbitrator follows existing substantive law and honors any claims of privilege recognized by law [4].\n"
+     ]
+    }
+   ],
+   "source": [
+    "query['_source'].append('shingles')\n",
+    "do_query(query)"
+   ]
+  }
+ ],
```
Comment (jonfritz) [inline notebooks/ndd_example.ipynb:180]: I think you need a cell explaining what changed in the next query, and what happened. This is a pretty complicated JSON, and most people will not know OpenSearch syntax.

**Review by jonfritz — COMMENTED** [review]
Very cool - a few suggestions to help a reader understand what they are seeing here. Most users will not know a ton about OpenSearch syntax.

File: notebooks/ndd_example.ipynb (hunk 1 above)
Comment (alexaryn) [reply notebooks/ndd_example.ipynb:8]: Seems like 4-6 hours on my laptop (Linux, 16 cores, 32GB).  Of 1911 PDFs, 17 failed to ingest for some reason.  The importer keeps trying, but the failures seem persistent.

**Comment by alexaryn** [comment]
> To get this into the containers, you need to copy it into the jupyter notebook, and probably make it work with both localhost:9200 and opensearch:9200.

@eric-anderson, I planned to include it in the examples directory that ships with the Jupyter container.  Do we still need to complicate it with localhost?

**Comment by eric-anderson** [comment]
> > To get this into the containers, you need to copy it into the jupyter notebook, and probably make it work with both localhost:9200 and opensearch:9200.
> 
> @eric-anderson, I planned to include it in the examples directory that ships with the Jupyter container. Do we still need to complicate it with localhost?

I'll leave it to @jonfritz for the final decision. I've definitely found it's useful sometimes to have the scripts run outside, and the detection logic is pretty easy.  You definitely need to copy it in to ship it with examples, and I'd say do that in this PR (it seems like the "same thing" to me).

File: notebooks/ndd_example.ipynb [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+    "\n",
+    "- All 1911 PDFs (~4 hours): `docker compose run sycamore_crawler_s3 aryn-public cccmad`\n",
+    "- 35 PDFs needed for demo (minutes): `docker compose run sycamore_crawler_s3 aryn-public cccmad-tiny`\n",
+    "\n",
+    "Note that for the full dataset, approximately 17 documents will fail to ingest for various reasons.  The Sycamore importer will keep retrying them, but the problems will persist.  It's not necessary for 100% of the documents to be ingested in order to run this example.  Once the importer queue shrinks to ~100 documents, it's OK to proceed.  Note that due to variations in ingestion, results may not exactly match what's described here."
+   ]
```
Comment (eric-anderson) [inline notebooks/ndd_example.ipynb:14]: I rewrote it slightly to add more newlines (made it much more readable to me). I'd like to break up step 3, but couldn't figure out how.  Results below:

```
This notebook shows the results of query-time near-duplicate detection (NDD).  It shows the same query with and without the duplicates.  

To use this notebook:
1. Follow [these instructions](https://sycamore.readthedocs.io/en/stable/welcome_to_sycamore/get_started.html) and start the Sycamore containers using `docker compose up`.
2. Ingest the college credit card marketing agreements data.  The documents come from [data.gov](https://catalog.data.gov/dataset/college-credit-card-marketing-agreements-data), but we have made them accessible via S3.  There are two ingestion commands to choose from, depending on how realistic or quick the demo should be:
    - All 1911 PDFs (~4 hours): `docker compose run sycamore_crawler_s3 aryn-public cccmad`
    - 35 PDFs needed for demo (minutes): `docker compose run sycamore_crawler_s3 aryn-public cccmad-tiny`
3. If you imported the full dataset, you can proceed once the importer queue shrinks to ~100 documents. It's not necessary for 100% of the documents to be ingested in order to run this example.   Due to variations in ingestion, results may not exactly match what's described here. Note that for the full dataset, approximately 17 documents will fail to ingest for various reasons.  The Sycamore importer will keep retrying them, but the problems will persist.
```

File: notebooks/ndd_example.ipynb (hunk 1 above)
Comment (jonfritz) [reply notebooks/ndd_example.ipynb]: Gotcha. Do we know how long the retrying period lasts? E.g. is it 4 hours of processing, 20 mins of flailing?

File: notebooks/ndd_example.ipynb [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+    "        res = json.loads(resp.text)\n",
+    "        return res['hits']['hits'][0]['_id']"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c94d8d48-a8b6-4704-a709-5cd022ed7343",
+   "metadata": {},
+   "source": [
+    "This next function performs the supplied query and prints out both the retrieved chunks and the AI-generated answer.  For clarity, the text chunks are truncated at 80 characters."
```
Comment (eric-anderson) [inline notebooks/ndd_example.ipynb:70]: `<br><hr>`

at the beginning of each of the markdown sections.

**Comment by jonfritz** [comment]
I think including it in examples would be fine. Either way.

**Review by jonfritz — APPROVED** [review]
LGTM. Thx!

**Review by eric-anderson — APPROVED** [review]

File: notebooks/ndd_example.ipynb (hunk 1 above)
Comment (alexaryn) [reply notebooks/ndd_example.ipynb]: The retrying continues forever, I think.

Comment (jonfritz) [reply notebooks/ndd_example.ipynb]: That seems like an issue. @eric-anderson we don't gracefully fail?

File: notebooks/ndd_example.ipynb (hunk 6 above)
Comment (alexaryn) [reply notebooks/ndd_example.ipynb:70]: It turns out <br><hr> messes up the backtick formatting.  <br> and --- works better.

**Pull request merged**